### DISCUSSION TITLE
Noob Networking problem on Windows with Rootless Podman 4.6.0 and Podman Desktop v1.2.1 and IPv4 connectivity

### DISCUSSION THREAD

**Discussion by reo-1963** [containers/podman#19575] [Q&A] [answered]
I am new to "podman" and new to running Linux. But I have been trying hard for the last 4 days to figure out as
much as I could on my own and limit the amount of "noob" questions. But I am stuck with a "noob" problem.
I'm trying to figure out how to get my Windows 10 PC's machine IPv4 address to connect to a Podman container.
Feels like I am close but I think I am missing something on the setting up of the WSL side of my PC.

I am running "rootless". And all the "podman" commands I run via a Windows CMD.EXE:
podman network create --ipv6 podman1
podman run -d -p 8080:80 --network podman1 --name no_poop quay.io/libpod/banner

I can run from CMD.EXE the following and see the "podman ascii art" appear:
   curl 127.0.0.1:8080
   curl [::1]:8080

What I am looking to run is my machine's IPv4 address (and see the "podman ascii art"):
    curl 172.12.12.96:8080
but all I get is:
   curl: (7) Failed to connect to 172.12.12.96 port 8080 after 2049 ms: Couldn't connect to server

If I use "Docker" and turn off all the "Podman" stuff, I can run the "swaggerapi/swagger-ui" on port 8080
and URL to "172.12.12.96:8080", I can see the standard petshop swagger page
so I am fairly confident I do not have a firewall issue.

I can't run the:
   podman network inspect podman1 | jq .[] > ~/.local/share/containers/storage/networks/podman.json
from WSL because the "podman" command is not recognized or installed there. It's only on Windows.
I don't have "jq" installed on my Windows machine.
I have piped the "podman network inspect podman1" command into a file and copied that file to:
    C:\Users\ROberuc\.local\share\containers\storage\networks\podman.json
but I have no idea how the "jq" would alter that command.

And I was able to install the "slirp4netns" with the
sudo apt-get install slirp4netns
but I am not certain what I do after this... 
Or how to check what the "iptables" are doing because any attempt to anything with that results 
'# iptables -S
iptables v1.8.4 (legacy): can't initialize iptables table `filter': Table does not exist (do you need to insmod?)
Perhaps iptables or your kernel needs to be upgraded.

Are there other diagnostics I can run for clues?
If there are other links in here that have covered this, please let me know.

Running in the Terminal of the container:
'# hostname
0e223ab05ca8

'# ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0@if4: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1500 qdisc noqueue state UP qlen 1000
    link/ether f6:b3:df:5f:35:db brd ff:ff:ff:ff:ff:ff
    inet 10.89.0.2/24 brd 10.89.0.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fdee:3784:f2:e33b::2/64 scope global 
       valid_lft forever preferred_lft forever
    inet6 fe80::f4b3:dfff:fe5f:35db/64 scope link 
       valid_lft forever preferred_lft forever

'# route
  Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
default         10.89.0.1       0.0.0.0         UG    100    0        0 eth0
10.89.0.0       *               255.255.255.0   U     0      0        0 eth0     

podman inspect no_poop
================================
[
     {
          "Id": "0e223ab05ca8d130c3b56fcd00257dd4a92888c0d1c9e583e90495add15d586d",
          "Created": "2023-08-09T14:17:40.759584983-04:00",
          "Path": "nginx",
          "Args": [
               "-g",
               "daemon off;"
          ],
          "State": {
               "OciVersion": "1.1.0-rc.3",
               "Status": "running",
               "Running": true,
               "Paused": false,
               "Restarting": false,
               "OOMKilled": false,
               "Dead": false,
               "Pid": 276,
               "ConmonPid": 274,
               "ExitCode": 0,
               "Error": "",
               "StartedAt": "2023-08-09T15:27:21.16661096-04:00",
               "FinishedAt": "0001-01-01T00:00:00Z",
               "Health": {
                    "Status": "",
                    "FailingStreak": 0,
                    "Log": null
               },
               "CgroupPath": "/user.slice/user-1000.slice/user@1000.service/app.slice/podman.service",
               "CheckpointedAt": "0001-01-01T00:00:00Z",
               "RestoredAt": "0001-01-01T00:00:00Z"
          },
          "Image": "5ba9aec95f0f3f2a3ff3f76dd5a571519f34884adc6a98b8ef99aadd671a1768",
          "ImageDigest": "sha256:ba6884a8b8c54a32d4b982f970ee839184abd9cf9c58edc160e20ca6e8a41f64",
          "ImageName": "quay.io/libpod/banner:latest",
          "Rootfs": "",
          "Pod": "",
          "ResolvConfPath": "/run/user/1000/containers/overlay-containers/0e223ab05ca8d130c3b56fcd00257dd4a92888c0d1c9e583e90495add15d586d/userdata/resolv.conf",
          "HostnamePath": "/run/user/1000/containers/overlay-containers/0e223ab05ca8d130c3b56fcd00257dd4a92888c0d1c9e583e90495add15d586d/userdata/hostname",
          "HostsPath": "/run/user/1000/containers/overlay-containers/0e223ab05ca8d130c3b56fcd00257dd4a92888c0d1c9e583e90495add15d586d/userdata/hosts",
          "StaticDir": "/home/user/.local/share/containers/storage/overlay-containers/0e223ab05ca8d130c3b56fcd00257dd4a92888c0d1c9e583e90495add15d586d/userdata",
          "OCIConfigPath": "/home/user/.local/share/containers/storage/overlay-containers/0e223ab05ca8d130c3b56fcd00257dd4a92888c0d1c9e583e90495add15d586d/userdata/config.json",
          "OCIRuntime": "crun",
          "ConmonPidFile": "/run/user/1000/containers/overlay-containers/0e223ab05ca8d130c3b56fcd00257dd4a92888c0d1c9e583e90495add15d586d/userdata/conmon.pid",
          "PidFile": "/run/user/1000/containers/overlay-containers/0e223ab05ca8d130c3b56fcd00257dd4a92888c0d1c9e583e90495add15d586d/userdata/pidfile",
          "Name": "no_poop",
          "RestartCount": 0,
          "Driver": "overlay",
          "MountLabel": "",
          "ProcessLabel": "",
          "AppArmorProfile": "",
          "EffectiveCaps": [
               "CAP_CHOWN",
               "CAP_DAC_OVERRIDE",
               "CAP_FOWNER",
               "CAP_FSETID",
               "CAP_KILL",
               "CAP_NET_BIND_SERVICE",
               "CAP_SETFCAP",
               "CAP_SETGID",
               "CAP_SETPCAP",
               "CAP_SETUID",
               "CAP_SYS_CHROOT"
          ],
          "BoundingCaps": [
               "CAP_CHOWN",
               "CAP_DAC_OVERRIDE",
               "CAP_FOWNER",
               "CAP_FSETID",
               "CAP_KILL",
               "CAP_NET_BIND_SERVICE",
               "CAP_SETFCAP",
               "CAP_SETGID",
               "CAP_SETPCAP",
               "CAP_SETUID",
               "CAP_SYS_CHROOT"
          ],
          "ExecIDs": [],
          "GraphDriver": {
               "Name": "overlay",
               "Data": {
                    "LowerDir": "/home/user/.local/share/containers/storage/overlay/fe2cbcae5c8a0b19183e5e08cc796d9007c49aeb3745ec60e1c2eef36ccb3a1a/diff:/home/user/.local/share/containers/storage/overlay/706fc2af1683c7112aa3ad1041c95e692d893d45c813641f99bad19f43c308f7/diff:/home/user/.local/share/containers/storage/overlay/26038dabf3bfc599c444ba3e5d803f448d0b2cfec58369ea7f059a65d3a04238/diff:/home/user/.local/share/containers/storage/overlay/24302eb7d9085da80f016e7e4ae55417e412fb7e0a8021e95e3b60c67cde557d/diff",
                    "MergedDir": "/home/user/.local/share/containers/storage/overlay/830bf27843b6272a8d953ccb1b7607e79277ff1a9d7e2f0c981aa205c780d648/merged",
                    "UpperDir": "/home/user/.local/share/containers/storage/overlay/830bf27843b6272a8d953ccb1b7607e79277ff1a9d7e2f0c981aa205c780d648/diff",
                    "WorkDir": "/home/user/.local/share/containers/storage/overlay/830bf27843b6272a8d953ccb1b7607e79277ff1a9d7e2f0c981aa205c780d648/work"
               }
          },
          "Mounts": [],
          "Dependencies": [],
          "NetworkSettings": {
               "EndpointID": "",
               "Gateway": "",
               "IPAddress": "",
               "IPPrefixLen": 0,
               "IPv6Gateway": "",
               "GlobalIPv6Address": "",
               "GlobalIPv6PrefixLen": 0,
               "MacAddress": "",
               "Bridge": "",
               "SandboxID": "",
               "HairpinMode": false,
               "LinkLocalIPv6Address": "",
               "LinkLocalIPv6PrefixLen": 0,
               "Ports": {
                    "80/tcp": [
                         {
                              "HostIp": "",
                              "HostPort": "8080"
                         }
                    ]
               },
               "SandboxKey": "/run/user/1000/netns/netns-df9b8cfa-7bed-659e-3668-25b29f641e19",
               "Networks": {
                    "podman1": {
                         "EndpointID": "",
                         "Gateway": "10.89.0.1",
                         "IPAddress": "10.89.0.2",
                         "IPPrefixLen": 24,
                         "IPv6Gateway": "fdee:3784:f2:e33b::1",
                         "GlobalIPv6Address": "fdee:3784:f2:e33b::2",
                         "GlobalIPv6PrefixLen": 64,
                         "MacAddress": "f6:b3:df:5f:35:db",
                         "NetworkID": "podman1",
                         "DriverOpts": null,
                         "IPAMConfig": null,
                         "Links": null,
                         "Aliases": [
                              "0e223ab05ca8"
                         ]
                    }
               }
          },
          "Namespace": "",
          "IsInfra": false,
          "IsService": false,
          "KubeExitCodePropagation": "invalid",
          "lockNumber": 0,
          "Config": {
               "Hostname": "0e223ab05ca8",
               "Domainname": "",
               "User": "",
               "AttachStdin": false,
               "AttachStdout": false,
               "AttachStderr": false,
               "Tty": false,
               "OpenStdin": false,
               "StdinOnce": false,
               "Env": [
                    "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
                    "TERM=xterm",
                    "container=podman",
                    "HOME=/root",
                    "HOSTNAME=0e223ab05ca8"
               ],
               "Cmd": [
                    "nginx",
                    "-g",
                    "daemon off;"
               ],
               "Image": "quay.io/libpod/banner:latest",
               "Volumes": null,
               "WorkingDir": "/",
               "Entrypoint": "",
               "OnBuild": null,
               "Labels": null,
               "Annotations": {
                    "io.container.manager": "libpod",
                    "org.opencontainers.image.stopSignal": "15"
               },
               "StopSignal": 15,
               "HealthcheckOnFailureAction": "none",
               "CreateCommand": [
                    "podman",
                    "run",
                    "-d",
                    "-p",
                    "8080:80",
                    "--network",
                    "podman1",
                    "--name",
                    "no_poop",
                    "quay.io/libpod/banner"
               ],
               "Umask": "0022",
               "Timeout": 0,
               "StopTimeout": 10,
               "Passwd": true,
               "sdNotifyMode": "container"
          },
          "HostConfig": {
               "Binds": [],
               "CgroupManager": "cgroupfs",
               "CgroupMode": "host",
               "ContainerIDFile": "",
               "LogConfig": {
                    "Type": "journald",
                    "Config": null,
                    "Path": "",
                    "Tag": "",
                    "Size": "0B"
               },
               "NetworkMode": "bridge",
               "PortBindings": {
                    "80/tcp": [
                         {
                              "HostIp": "",
                              "HostPort": "8080"
                         }
                    ]
               },
               "RestartPolicy": {
                    "Name": "",
                    "MaximumRetryCount": 0
               },
               "AutoRemove": false,
               "VolumeDriver": "",
               "VolumesFrom": null,
               "CapAdd": [],
               "CapDrop": [],
               "Dns": [],
               "DnsOptions": [],
               "DnsSearch": [],
               "ExtraHosts": [],
               "GroupAdd": [],
               "IpcMode": "shareable",
               "Cgroup": "",
               "Cgroups": "default",
               "Links": null,
               "OomScoreAdj": 0,
               "PidMode": "private",
               "Privileged": false,
               "PublishAllPorts": false,
               "ReadonlyRootfs": false,
               "SecurityOpt": [],
               "Tmpfs": {},
               "UTSMode": "private",
               "UsernsMode": "",
               "ShmSize": 65536000,
               "Runtime": "oci",
               "ConsoleSize": [
                    0,
                    0
               ],
               "Isolation": "",
               "CpuShares": 0,
               "Memory": 0,
               "NanoCpus": 0,
               "CgroupParent": "",
               "BlkioWeight": 0,
               "BlkioWeightDevice": null,
               "BlkioDeviceReadBps": null,
               "BlkioDeviceWriteBps": null,
               "BlkioDeviceReadIOps": null,
               "BlkioDeviceWriteIOps": null,
               "CpuPeriod": 0,
               "CpuQuota": 0,
               "CpuRealtimePeriod": 0,
               "CpuRealtimeRuntime": 0,
               "CpusetCpus": "",
               "CpusetMems": "",
               "Devices": [],
               "DiskQuota": 0,
               "KernelMemory": 0,
               "MemoryReservation": 0,
               "MemorySwap": 0,
               "MemorySwappiness": 0,
               "OomKillDisable": false,
               "PidsLimit": 0,
               "Ulimits": [
                    {
                         "Name": "RLIMIT_NPROC",
                         "Soft": 32768,
                         "Hard": 32768
                    },
                    {
                         "Name": "RLIMIT_NOFILE",
                         "Soft": 1048576,
                         "Hard": 1048576
                    }
               ],
               "CpuCount": 0,
               "CpuPercent": 0,
               "IOMaximumIOps": 0,
               "IOMaximumBandwidth": 0,
               "CgroupConf": null
          }
     }
]

podman network inspect podman1
===============================
[
     {
          "name": "podman1",
          "id": "3dd34028cc879877b62a6ab4f4b52e83c8d795a86ff0d50aea7748f60c4d87fd",
          "driver": "bridge",
          "network_interface": "podman1",
          "created": "2023-08-09T13:41:26.20844695-04:00",
          "subnets": [
               {
                    "subnet": "10.89.0.0/24",
                    "gateway": "10.89.0.1"
               },
               {
                    "subnet": "fdee:3784:f2:e33b::/64",
                    "gateway": "fdee:3784:f2:e33b::1"
               }
          ],
          "ipv6_enabled": true,
          "internal": false,
          "dns_enabled": true,
          "ipam_options": {
               "driver": "host-local"
          }
     }
]

podman info --debug
===========================
host:
  arch: amd64
  buildahVersion: 1.31.0
  cgroupControllers: []
  cgroupManager: cgroupfs
  cgroupVersion: v1
  conmon:
    package: conmon-2.1.7-2.fc37.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.1.7, commit: '
  cpuUtilization:
    idlePercent: 99.94
    systemPercent: 0.04
    userPercent: 0.02
  cpus: 16
  databaseBackend: boltdb
  distribution:
    distribution: fedora
    variant: container
    version: "37"
  eventLogger: journald
  freeLocks: 2047
  hostname: ROberuc-3650
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 524288
      size: 65536
    uidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 524288
      size: 65536
  kernel: 5.15.90.1-microsoft-standard-WSL2
  linkmode: dynamic
  logDriver: journald
  memFree: 32672624640
  memTotal: 33514868736
  networkBackend: netavark
  networkBackendInfo:
    backend: netavark
    dns:
      package: aardvark-dns-1.7.0-1.fc37.x86_64
      path: /usr/libexec/podman/aardvark-dns
      version: aardvark-dns 1.7.0
    package: netavark-1.7.0-1.fc37.x86_64
    path: /usr/libexec/podman/netavark
    version: netavark 1.7.0
  ociRuntime:
    name: crun
    package: crun-1.8.6-1.fc37.x86_64
    path: /usr/bin/crun
    version: |-
      crun version 1.8.6
      commit: 73f759f4a39769f60990e7d225f561b4f4f06bcf
      rundir: /run/user/1000/crun
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +CRIU +LIBKRUN +WASM:wasmedge +YAJL
  os: linux
  pasta:
    executable: /usr/bin/pasta
    package: passt-0^20230625.g32660ce-1.fc37.x86_64
    version: |
      pasta 0^20230625.g32660ce-1.fc37.x86_64
      Copyright Red Hat
      GNU Affero GPL version 3 or later <https://www.gnu.org/licenses/agpl-3.0.html>
      This is free software: you are free to change and redistribute it.
      There is NO WARRANTY, to the extent permitted by law.
  remoteSocket:
    exists: true
    path: /run/user/1000/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: true
    seccompEnabled: true
    seccompProfilePath: /usr/share/containers/seccomp.json
    selinuxEnabled: false
  serviceIsRemote: true
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: slirp4netns-1.2.0-8.fc37.x86_64
    version: |-
      slirp4netns version 1.2.0
      commit: 656041d45cfca7a4176f6b7eed9e4fe6c11e8383
      libslirp: 4.7.0
      SLIRP_CONFIG_VERSION_MAX: 4
      libseccomp: 2.5.3
  swapFree: 8589934592
  swapTotal: 8589934592
  uptime: 0h 53m 39.00s
plugins:
  authorization: null
  log:
  - k8s-file
  - none
  - passthrough
  - journald
  network:
  - bridge
  - macvlan
  - ipvlan
  volume:
  - local
registries:
  search:
  - docker.io
store:
  configFile: /home/user/.config/containers/storage.conf
  containerStore:
    number: 1
    paused: 0
    running: 1
    stopped: 0
  graphDriverName: overlay
  graphOptions: {}
  graphRoot: /home/user/.local/share/containers/storage
  graphRootAllocated: 269427478528
  graphRootUsed: 1160273920
  graphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "false"
  imageCopyTmpDir: /var/tmp
  imageStore:
    number: 1
  runRoot: /run/user/1000/containers
  transientStore: false
  volumePath: /home/user/.local/share/containers/storage/volumes
version:
  APIVersion: 4.6.0
  Built: 1689942107
  BuiltTime: Fri Jul 21 08:21:47 2023
  GitCommit: ""
  GoVersion: go1.19.10
  Os: linux
  OsArch: linux/amd64
  Version: 4.6.0

**Comment by Luap99** [comment, marked as answer]
WSL2 does not allow such connections, see https://learn.microsoft.com/en-us/windows/wsl/networking#accessing-a-wsl-2-distribution-from-your-local-area-network-lan

However podman v4.6 added a `--user-mode-networking` to `podman machine init` that should make it work AFAIK.

**Comment by reo-1963** [comment]
Thank you!

DOS> netsh interface portproxy add v4tov4 listenport=8081 listenaddress=0.0.0.0 connectport=8080 connectaddress=127.0.0.1

did the trick! And then "curl 172.17.12.96:8081" works fine. For me and local work colleagues on PC's across our network.
A colleague recommended the "netsh interface portproxy" earlier but it did not work...
Part of getting it to work is setting up a better "podman network create --ipv6" (which adds a few things more than the default "podman" network has) and then using that "podman1" network with the run.
And if anyone else has this problem, test using "curl" and not your web browser because the browser caching will mess you up.
And be warned about setting the "listenport=8080". It worked once... until I stopped the container and restarted it.
I think "portproxy" got confused about the circular nature of the piping. So use 2 different ports with it.
